File: 진태/주사위.py
def Down():
    temp = position["top"];
    position["top"] = position["back"];
    position["back"] = position["bottom"];
    position["bottom"] = position["front"];
    position["front"] = temp;

def Up():
    temp = position["top"];
    position["top"] = position["front"];
    position["front"] = position["bottom"];
    position["bottom"] = position["back"];
    position["back"] = temp;

def Left():
    temp = position["top"];
    position["top"] = position["right"];
    position["right"] = position["bottom"];
    position["bottom"] = position["left"];
    position["left"] = temp;

def Right():
    temp = position["top"];
    position["top"] = position["left"];
    position["left"] = position["bottom"];
    position["bottom"] = position["right"];
    position["right"] = temp;


def IsValid(d):
    nx = xy[0] + dx[d];
    ny = xy[1] + dy[d];

    if (nx < 0 or nx >= n):
        return True;
    if (ny < 0 or ny >= m):
        return True;

    return False;

def Check():
    if (map[xy[0]][xy[1]] == 0):
        map[xy[0]][xy[1]] = dice[position["bottom"]];
    else:
        dice[position["bottom"]] = map[xy[0]][xy[1]];
        map[xy[0]][xy[1]] = 0;

    print(dice[position["top"]]);

# 기본 입력
n, m, x, y, k = input().split();
n = int(n);
m = int(m);
x = int(x);
y = int(y);
xy = [x, y];
k = int(k);

# 맵과 주사위 설정, 주사위는 indexing이 편하게 7개 설정
map = [[0 for cols in range(m)]for rows in range(n)];
dice = [0] * 7;

# 맵 입력
for i in range(n):
    tmp = input().split();
    for j in range(m):
        map[i][j] = int(tmp[j]);

# 이동 명령 입력
go = input().split();

# 주사위 기본 위치 입력
position = {"dump":0, "top":1, "back":2, "right":3, "left":4, "front":5, "bottom":6};

# 방향함수도 편의성을 위해 index 0은 사용 X
dx = [0, 0, 0, -1, 1];
dy = [0, 1, -1, 0, 0];

# 주사위 이동 시작
for i in range(k):
    direction = int(go[i]);
    if (direction == 1):
        if(IsValid(direction) == False):
            xy[0] = xy[0] + dx[direction];
            xy[1] = xy[1] + dy[direction];
            Right();
            Check();
        else:
            continue;
    elif (direction == 2):
        if(IsValid(direction) == False):
            xy[0] = xy[0] + dx[direction];
            xy[1] = xy[1] + dy[direction];
            Left();
            Check();
        else:
            continue;
    elif (direction == 3):
        if(IsValid(direction) == False):
            xy[0] = xy[0] + dx[direction];
            xy[1] = xy[1] + dy[direction];
            Up();
            Check();
        else:
            continue;
    else:
        if(IsValid(direction) == False):
            xy[0] = xy[0] + dx[direction];
            xy[1] = xy[1] + dy[direction];
            Down();
            Check();
        else:
            continue;

File: 진태/주사위.cpp
#include <cstdio>

using namespace std;

int map[20][20];
int n, m, x, y, k;
int go;

int dx[] = {0, 0, 0, -1, 1 };
int dy[] = {0, 1, -1, 0, 0 };

int left = 4;
int right = 3;
int top = 1;
int bottom = 6;
int front = 5;
int back = 2;

int dice[7];

bool IsValid(int d) {
	int nx = x + dx[d];
	int ny = y + dy[d];

	if (0 > nx || nx >= n)
		return false;
	if (0 > ny || ny >= m)
		return false;

	return true;
}

void Rolling(int d) {
	int temp;

	switch (d) {
	case 1:
		temp = top;
		top = left;
		left = bottom;
		bottom = right;
		right = temp;
		break;
	case 2:
		temp = top;
		top = right;
		right = bottom;
		bottom = left;
		left = temp;
		break;
	case 3:
		temp = top;
		top = front;
		front = bottom;
		bottom = back;
		back = temp;
		break;
	case 4:
		temp = top;
		top = back;
		back = bottom;
		bottom = front;
		front = temp;
	}
}

void Move(int direction) {
	if (IsValid(direction) == true) {
		x += dx[direction];
		y += dy[direction];

		Rolling(direction);

		if (map[x][y] == 0) {
			map[x][y] = dice[bottom];
		}
		else {
			dice[bottom] = map[x][y];
			map[x][y] = 0;
		}

		printf("%d\n", dice[top]);
	}
}

int main() {
	scanf("%d %d %d %d %d", &n, &m, &x, &y, &k);

	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			scanf("%d", &map[i][j]);

	for (int i = 0; i < k; i++) {
		scanf("%d", &go);

		Move(go);
	}

	return 0;
}